core/arch/arm/mm/mobj_ffa.c

```diff
@@ -0,0 +1,562 @@
+				     get_page_count(mf));
+		tee_mm_free(mf->mm);
+		mf->mm = NULL;
+	}
+}
+
+TEE_Result mobj_ffa_unregister_by_cookie(uint64_t cookie)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct mobj_ffa *mf = NULL;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+	mf = find_in_list(&shm_head, cmp_cookie, cookie);
+	/*
+	 * If the mobj is found here it's still active and cannot be
+	 * unregistered.
+	 */
+	if (mf) {
+		DMSG("cookie %#"PRIx64" busy refc %u",
+		     cookie, refcount_val(&mf->mobj.refc));
+		res = TEE_ERROR_BUSY;
+		goto out;
+	}
+	mf = find_in_list(&shm_inactive_head, cmp_cookie, cookie);
+	/*
+	 * If the mobj isn't found or if it already have been unregistered.
+	 */
+	if (!mf || mf->unregistered_by_cookie) {
+		res = TEE_ERROR_ITEM_NOT_FOUND;
+		goto out;
+	}
+	mf->unregistered_by_cookie = true;
+	res = TEE_SUCCESS;
+
+out:
+	cpu_spin_unlock_xrestore(&shm_lock, exceptions);
+	return res;
+}
+
+#ifdef CFG_CORE_SEL1_SPMC
+TEE_Result mobj_ffa_sel1_spmc_reclaim(uint64_t cookie)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct mobj_ffa *mf = NULL;
+	uint32_t exceptions = 0;
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+	mf = find_in_list(&shm_head, cmp_cookie, cookie);
+	/*
+	 * If the mobj is found here it's still active and cannot be
+	 * reclaimed.
+	 */
+	if (mf) {
+		DMSG("cookie %#"PRIx64" busy refc %u",
+		     cookie, refcount_val(&mf->mobj.refc));
+		res = TEE_ERROR_BUSY;
+		goto out;
+	}
+
+	mf = find_in_list(&shm_inactive_head, cmp_cookie, cookie);
+	/* If the mobj isn't found */
+	if (!mf) {
+		res = TEE_ERROR_ITEM_NOT_FOUND;
+		goto out;
+	}
+	/*
+	 * If the mobj has been registered via mobj_ffa_get_by_cookie()
+	 * but not unregistered yet with mobj_ffa_unregister_by_cookie().
+	 */
+	if (mf->registered_by_cookie && !mf->unregistered_by_cookie) {
+		DMSG("cookie %#"PRIx64" busy", cookie);
+		res = TEE_ERROR_BUSY;
+		goto out;
+	}
+
+	if (!pop_from_list(&shm_inactive_head, cmp_ptr, (vaddr_t)mf))
+		panic();
+	res = TEE_SUCCESS;
+out:
+	cpu_spin_unlock_xrestore(&shm_lock, exceptions);
+	if (!res)
+		mobj_ffa_sel1_spmc_delete(mf);
+	return res;
+}
+#endif /*CFG_CORE_SEL1_SPMC*/
+
+struct mobj *mobj_ffa_get_by_cookie(uint64_t cookie, unsigned int internal_offs)
+{
+	struct mobj_ffa *mf = NULL;
+	uint32_t exceptions = 0;
+
+	if (internal_offs >= SMALL_PAGE_SIZE)
+		return NULL;
+
+	exceptions = cpu_spin_lock_xsave(&shm_lock);
+
+	mf = find_in_list(&shm_head, cmp_cookie, cookie);
+	if (mf) {
+		if (mf->page_offset == internal_offs) {
+			if (!refcount_inc(&mf->mobj.refc)) {
+				/*
+				 * If refcount is 0 some other thread has
+				 * called mobj_put() on this reached 0 and
+				 * before ffa_inactivate() got the lock we
+				 * found it. Let's reinitialize it.
+				 */
+				refcount_set(&mf->mobj.refc, 1);
+			}
+			DMSG("cookie %#"PRIx64" active: refc %d",
+			     cookie, refcount_val(&mf->mobj.refc));
+		} else {
+			EMSG("cookie %#"PRIx64" mismatching internal_offs got %#"PRIx16" expected %#x",
```

Comment:
Move traces should be outside the spinlock xsave?

Reply:
Why? What's the concern?

Reply:
ok as is